### PR TITLE
Digital twin

### PR DESCRIPTION
**Pull request by Bravo555**
## TODO

- [x] adding entities
- [x] listing all entities
- [x] get the main device
- [x] get/set an entity struct for an entity identifier `+/+/+/+`
- [x] list the service running on a device
- [x] list child devices of a device or child device
- [x] provide a channel where updates are notified (`update()` returns a list of affected entities)
- [x] create empty twin with a function to derive a name from an (identifier, entity) pair
- [x] derive a name for each entity (indexing this names)
- [x] get an entity identifier from and entity name
- [x] ~~make `entity_store` use `entity` API~~ not a blocker, can be done later

## Proposed changes

Add `entity` and `entity_store` modules that represent entities and allow keeping track of registered entities, respectively.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/2081

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/c8y_mapper_ext/src/actor.rs [hunk 1]
```diff
@@ -110,6 +133,63 @@ impl C8yMapperActor {
...
+    ///
+    /// It returns MQTT register messages for the given entities to be published
+    /// by the mapper, so other components can also be aware of a new device
+    /// being registered.
+    fn auto_register_entity(&mut self, entity_id: &str) -> Vec<Message> {
+        let mut register_messages = vec![];
+        if let Some(("te", topic)) = entity_id.split_once('/') {
+            let (device_id, service_id) = match topic.split('/').collect::<Vec<&str>>()[..] {
+                ["device", device_id, "service", service_id, ..] => (device_id, Some(service_id)),
+                ["device", device_id, ..] => (device_id, None),
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/actor.rs:150]: I would discard any message that strictly follow the default scheme.

```suggestion
                ["device", device_id, "", ""] => (device_id, None),
```

File: crates/extensions/c8y_mapper_ext/src/actor.rs [hunk 2]
```diff
@@ -110,6 +133,63 @@ impl C8yMapperActor {
...
 
+    /// Performs auto-registration process for an entity under a given
+    /// identifier.
+    ///
+    /// If an entity is a service, its device is also auto-registered if it's
+    /// not already registered.
+    ///
+    /// It returns MQTT register messages for the given entities to be published
+    /// by the mapper, so other components can also be aware of a new device
+    /// being registered.
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/actor.rs:144]: This is a good idea. But I would be cautious. Indeed, the measurement/alarm/event must not be sent *before* the device is actually registered on C8Y. But, the current code jumps into the translation.

**Review by didier-wenzek — COMMENTED** [review]
In a good shape! Thanks

File: crates/core/tedge_api/src/entity_store.rs [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+/// assert_eq!(entity_store.main_device(), Some("te/device/main//"));
+/// ```
+#[derive(Debug, Clone)]
+pub enum EntityStore {
+    Uninitialised,
+    Initialised {
+        main_device: EntityId,
+        entities: HashMap<EntityId, EntityMetadata>,
+    },
+}
```
[the remark this answers is not in the document]
Comment (Bravo555) [reply crates/core/tedge_api/src/entity_store.rs:57]: I talked with @reubenmiller about whether the main device also has to be registered, and whether or not it can be auto-registered, and I ended up rewriting this once again under the assumption that when the entity store is created, the main device is already known.

File: crates/core/tedge_api/src/entity_store.rs [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+
+                Ok(Vec::new())
+            }
+
+            EntityStore::Initialised {
+                entities,
+                main_device,
+            } => {
+                if message.r#type == "device" {
+                    return Err(Error::MainDeviceAlreadyRegistered);
```
[the remark this answers is not in the document]
Comment (Bravo555) [reply crates/core/tedge_api/src/entity_store.rs]: fixed

File: crates/core/tedge_api/src/entity_store.rs [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+#[derive(Debug, Clone)]
+pub struct EntityStore {
+    main_device: EntityId,
+    entities: HashMap<EntityId, EntityMetadata>,
+    external_id_index: HashMap<String, EntityId>,
+}
+
+impl EntityStore {
+    /// Creates a new entity store with a given main device.
+    pub fn with_main_device(main_device: EntityRegistrationMessage) -> Option<Self> {
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:69]: This is okay for a first version  but will not work in a setting where the main-device topic is defined by the user using a registration message.

When the mapper starts, the id of the main device is known (This is the CN of its certificate) but not its topic (unless we are using the default topic scheme where the main device is attached to the well-known topic `te/device/main//`).

File: crates/core/tedge_api/src/entity_store.rs [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+    /// # Examples
+    /// - `device/main//` => `DEVICE_COMMON_NAME`
+    /// - `device/child001//` => `DEVICE_COMMON_NAME:device:child001`
+    /// - `device/child001/service/service001` => `DEVICE_COMMON_NAME:device:child001:service:service001`
+    /// - `factory01/hallA/packaging/belt001` => `DEVICE_COMMON_NAME:factory01:hallA:packaging:belt001`
+    fn external_id(&self, entity_topic: EntityIdRef) -> String {
+        let main_device_external_id = &self.get(&self.main_device).unwrap().external_id;
+        let external_id_suffix = entity_topic.replace('/', ":");
+        let external_id_suffix = external_id_suffix.trim_matches(':');
+        format!("{main_device_external_id}:{external_id_suffix}")
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:202]: No. For the main device the answer is `DEVICE_COMMON_NAME` and not `DEVICE_COMMON_NAME:device:main`

File: crates/core/tedge_api/src/entity_store.rs [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+    /// The external id is generated by prefixing the id with main device name
+    /// (device_common_name) and then appending the MQTT entity topic with `/`
+    ///  characters replaced by `:`.
+    ///
+    /// # Examples
+    /// - `device/main//` => `DEVICE_COMMON_NAME`
+    /// - `device/child001//` => `DEVICE_COMMON_NAME:device:child001`
+    /// - `device/child001/service/service001` => `DEVICE_COMMON_NAME:device:child001:service:service001`
+    /// - `factory01/hallA/packaging/belt001` => `DEVICE_COMMON_NAME:factory01:hallA:packaging:belt001`
+    fn external_id(&self, entity_topic: EntityIdRef) -> String {
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:198]: A better name for this function is `derive_external_id` as this is computation used when the registration message doesn't provide an id.

File: crates/core/tedge_api/src/entity_store.rs [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+        // parent device is affected if new device is its child
+        if let Some(parent) = &parent {
+            if !self.entities.contains_key(parent) {
+                return Err(Error::NoParent(parent.clone().into_boxed_str()));
+            }
+
+            affected_entities.push(parent.clone());
+        }
+
+        let external_id = self.external_id(&message.mqtt_id);
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:160]: If the message provides an `external_id` this one must be used.

```suggestion
        let external_id = message.external_id.unwrap_or_else(||self.external_id(&message.mqtt_id));
```

File: crates/core/tedge_api/src/entity_store.rs [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EntityMetadata {
+    parent: Option<EntityId>,
+    r#type: String,
+    external_id: String,
+    other: serde_json::Value,
+}
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:212]: I would consider to use an enum here, ruling out the cases that makes no sense.

```suggestion
#[derive(Debug, Clone)]
pub struct EntityMetadata {
    r#type: EntityType,
    other: serde_json::Value,
}

pub enum EntityType {
    MainDevice{ device_id: String },
    ChildDevice( device_id: String, parent: EntityId },
    Service( service_id: String, device: EntityId },
}
```

**Review by didier-wenzek — COMMENTED** [review]
We need to come with better names for the misc kind of identifier with have.
- We use in turn `identifier`, `entity_id`, `mqtt_id`, `external_id`,`device_id` for related but different things.
- `external_id` is confusing (my bad  as I proposed it) as this is used by the c8y mapper for the internal id given by c8y. A better name would be in fact `device_id`: for the main device this is the CN of its certificate , for a child device this is name assigned by the user or derived from the parent device id and the child identifier.
- You the term `mqtt_id` somewhere in this PR. I like it. This might be less confusing than `device_identifier` which rhymes with `device_id`. Other alternatives could be `device_topic` or `entity_topic`.

File: crates/core/tedge_api/src/entity_store.rs (hunk 9 above)
Comment (Bravo555) [reply crates/core/tedge_api/src/entity_store.rs]: Because `device_id` and `service_id` both refer to the external id, and in the case of child devices and services `parent` and `device` mean basically the same thing, I've opted to make the entity type just a plain enum.

File: crates/core/tedge_api/src/entity.rs [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ChannelCategory {
+    Measurement,
+    Event,
+    Alarm,
+    Command,
+}
```
[the remark this answers is not in the document]
Comment (Bravo555) [reply crates/core/tedge_api/src/entity.rs:193]: In this case, with all the `r#type`s being inside enum variants, it makes pattern matching on the channel harder. We'd be forced to match on every variant to extract the type, which is basically the same thing.
Perhaps the same type should not be used by multiple categories, so the set of possible values might be different for each variant, but the suggestion does not prevent that while making it a bit harder to match in the general case.

I'm also not very familiar with the idea of what's the impact of the `/meta` suffix and I'd like to keep it simple for now. I've aimed to make it a minimal wrapping over MQTT scheme, adding typing to very basic stuff. I'd rather see how this basic shape fits and then revisit this.

**Comment by Bravo555** [comment]
All the tests should pass now, will squash after positive review

File: crates/common/tedge_config_macros/Cargo.toml [hunk 11]
```diff
@@ -15,7 +15,7 @@ once_cell = "1.17"
 serde = "1.0"
 tedge_config_macros-macro = { path = "macro" }
 thiserror = "1.0"
+tracing = "0.1"
 url = "2.3"
 
 [dev-dependencies]
-tracing = "0.1"
```
Comment (didier-wenzek) [inline crates/common/tedge_config_macros/Cargo.toml:21]: If not moved into another PR this change must at least be extracted in a specific commit fixing `cargo doc`.

**Review by didier-wenzek — COMMENTED** [review]
I'm okay to merge this PR despite some discussions still open. This work is expected to unblock the work on the c8y mappers (to interact with  tedge-log-plugin and to use the new MQTT topic scheme).

We still need to discuss:
- [Channel type](https://github.com/thin-edge/thin-edge.io/pull/2122#discussion_r1293640062)
- [EntityType](https://github.com/thin-edge/thin-edge.io/pull/2122#discussion_r1293640062)
- [Naming](https://github.com/thin-edge/thin-edge.io/pull/2122#pullrequestreview-1577115611)

@Bravo555 can you clean the gitlog history squashing the fixup commits.

File: crates/common/tedge_config_macros/Cargo.toml (hunk 11 above)
Comment (Bravo555) [reply crates/common/tedge_config_macros/Cargo.toml]: extracted into #2160

**Review by didier-wenzek — APPROVED** [review]
Approved
